pipeline-model-api/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/ast/ModelASTStage.java

```diff
@@ -102,31 +116,37 @@ public String toGroovy() {
         if (environment != null) {
             result.append(environment.toGroovy());
         }
-        result.append("steps {\n");
-        if (branches.size() > 1) {
-            result.append("parallel(");
-            boolean first = true;
-            for (ModelASTBranch branch: branches) {
-                if (first) {
-                    first = false;
-                } else {
-                    result.append(',');
+        if (branches.isEmpty() && parallelStages != null) {
+            result.append("parallelStages {\n");
```

Comment:
And in that last case, why not allow all stages to potentially be run in parallel, even the outer ones :)